src/worker.cc

```diff
@@ -0,0 +1,694 @@
+      keep_alive_(keep_alive),
+      event_loop_(event_loop),
+      owner_notifications_(owner_event_loop(),
+                           OwnerNotificationCallback,
+                           this),
+      worker_notifications_(worker_event_loop(),
+                            WorkerNotificationCallback,
+                            this),
+      argc_(argc),
+      argv_(argv),
+      exec_argc_(exec_argc),
+      exec_argv_(exec_argv) {
+  CHECK_NE(event_loop, nullptr);
+  CHECK_NE(event_loop, uv_default_loop());
+  CHECK_CALLED_FROM_OWNER(this);
+  node::Wrap(owner_wrapper, this);
+
+  owner_env->AddSubWorkerContext(this);
+
+  CHECK_EQ(uv_mutex_init(termination_mutex()), 0);
+  CHECK_EQ(uv_mutex_init(api_mutex()), 0);
+  CHECK_EQ(uv_mutex_init(initialization_mutex()), 0);
+  CHECK_EQ(uv_mutex_init(to_owner_messages_mutex()), 0);
+  CHECK_EQ(uv_mutex_init(to_worker_messages_mutex()), 0);
+}
+
+void WorkerContext::New(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  HandleScope scope(env->isolate());
+  CHECK(args[0]->IsString());
+  CHECK(args[1]->IsObject());
+  Local<String> entry_module_path = args[0].As<String>();
+  bool keep_alive =
+      args[1].As<Object>()->Get(env->keep_alive_string())->BooleanValue();
+
+  uv_loop_t* worker_event_loop = static_cast<uv_loop_t*>(
+      malloc(sizeof(uv_loop_t)));
+  int err = uv_loop_init(worker_event_loop);
+  if (err != 0) {
+    free(worker_event_loop);
+    return env->ThrowUVException(err, "uv_loop_init");
+  }
+
+  // Heap allocated cause this stack frame can be gone by the time the worker
+  // needs them.
+  int argc = 2;
+  char** argv = new char*[argc];
+  // TODO(petkaantonov) should use correct executable name here
+  argv[0] = const_cast<char*>("node");
+  argv[1] = ToUtf8Value(env->isolate(), entry_module_path);
+  // TODO(petkaantonov) should use the process's real exec args
+  int exec_argc = 0;
+  char** exec_argv = new char*[exec_argc];
+
+  WorkerContext* context =
+      new WorkerContext(env,
+                        args.This(),
+                        worker_event_loop,
+                        argc,
+                        const_cast<const char**>(argv),
+                        exec_argc,
+                        const_cast<const char**>(exec_argv),
+                        keep_alive);
+
+  err = uv_thread_create(context->worker_thread(), RunWorkerInstance, context);
+
+  if (err != 0) {
+    context->worker_notifications()->Dispose();
+    context->Dispose();
+    CHECK_EQ(uv_loop_close(worker_event_loop), 0);
+    free(worker_event_loop);
+    return env->ThrowUVException(err, "uv_thread_create");
+  }
+}
+
+void WorkerContext::PostMessageToOwner(WorkerMessage* message) {
+  if (termination_kind() != NONE) {
+    delete message;
+    return;
+  }
+
+  if (!to_owner_messages_primary_.PushBack(message)) {
+    ScopedLock::Mutex lock(to_owner_messages_mutex());
+    to_owner_messages_backup_.PushBack(message);
```

Comment:
The primary queue is already such a queue, this queue is only used when that queue is full. Other thing that can be done in this situation is blocking until the primary queue has free slot.